source/falaise/snemo/datamodels/sim_tracker_digi_hit.cc

```diff
@@ -5,239 +5,321 @@
-  if (has_cathode_t6()) {
-    out_ << _cathode_t6_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::inherit_tag(inherit_)
-       << "Validity : " << (is_valid() ? "<yes>" : "<no>") << std::endl;
-
-  return;
-}
-
-}  // end of namespace datamodel
+  namespace datamodel {
+
+    // Serial tag :
+    DATATOOLS_SERIALIZATION_SERIAL_TAG_IMPLEMENTATION(sim_tracker_digi_hit,
+						      "snemo::datamodel::sim_tracker_digi_hit")
+
+    sim_tracker_digi_hit::sim_tracker_digi_hit()
+    {
+      return;
+    }
+
+    sim_tracker_digi_hit::~sim_tracker_digi_hit()
+    {
+      return;
+    }
+
+    bool sim_tracker_digi_hit::is_valid() const
+    {
+      if (!has_hit_id())
+	{
+	  return false;
+	}
+      if (!has_geom_id())
+	{
+	  return false;
+	}
+      if (!has_anode_R0())
+	{
+	  return false;
+	}
+      return true;
+    }
+
+    void sim_tracker_digi_hit::reset()
+    {
+      reset_anode_R0();
+      reset_anode_R1();
+      reset_anode_R2();
+      reset_anode_R3();
+      reset_anode_R4();
+      reset_cathode_R5();
+      reset_cathode_R6();
+      return;
+    }
+
+    bool sim_tracker_digi_hit::has_elec_id() const
+    {
+      return _elec_id_.is_valid();
+    }
+
+    void sim_tracker_digi_hit::invalidate_elec_id()
+    {
+      _elec_id_.invalidate();
+      return;
+    }
+
+    const geomtools::geom_id & sim_tracker_digi_hit::get_elec_id() const
+    {
+      return _elec_id_;
+    }
+
+    geomtools::geom_id & sim_tracker_digi_hit::grab_elec_id()
+    {
+      return _elec_id_;
+    }
+
+    void sim_tracker_digi_hit::set_elec_id(const geomtools::geom_id & a_gid)
+    {
+      _elec_id_ = a_gid;
+      return;
+    }
+
+    bool sim_tracker_digi_hit::has_anode_R0() const
+    {
+      return (_store & STORE_ANODE_R0);
+    }
+
+    bool sim_tracker_digi_hit::has_anode_R1() const
+    {
+      return (_store & STORE_ANODE_R1);
+    }
+
+    bool sim_tracker_digi_hit::has_anode_R2() const
+    {
+      return (_store & STORE_ANODE_R3);
+    }
+
+    bool sim_tracker_digi_hit::has_anode_R3() const
+    {
+      return (_store & STORE_ANODE_R3);
+    }
+
+    bool sim_tracker_digi_hit::has_anode_R4() const
+    {
+      return (_store & STORE_ANODE_R4);
+    }
+
+    bool sim_tracker_digi_hit::has_cathode_R5() const
+    {
+      return (_store & STORE_CATHODE_R5);
+    }
+
+    bool sim_tracker_digi_hit::has_cathode_R6() const
+    {
+      return (_store & STORE_CATHODE_R6);
+    }
+
+    void sim_tracker_digi_hit::set_anode_R0(int64_t t_)
+    {
+      _store_set(STORE_ANODE_R0);
+      _anode_R0_ = t_;
+      return;
+    }
+
+    void sim_tracker_digi_hit::reset_anode_R0()
+    {
+      _store_unset(STORE_ANODE_R0);
+      _anode_R0_ = INVALID_TIME;
+      return;
+    }
+
+    int64_t sim_tracker_digi_hit::get_anode_R0() const
+    {
+      DT_THROW_IF(!has_anode_R0(), std::logic_error, "Missing anode t0");
+      return _anode_R0_;
+    }
+
```

Comment:
I'm confused by this "store mask" pattern for each timestamp. The class provides an `INVALID_TIME` constant used as a marker of an unmeasured/bad value, why not just use this and let the client of the data decide what to do? The `has_XXX` interfaces can then go (same for the `reset`) and, the `get_/set_` methods are simplified.

I don't think the `get_...` should they and throw exceptions. The presence of invalid data isn't an "execeptional" case here.

Reply:
You're right, the value should be stored even if it's an invalid time. The invalid time is an information.
And it will garanty that there are 7 times recorded even if there is no data.  

I'll remove the DT throw too because we want to access to timestamps even if they are invalid